src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/TimeoutsTest.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    /**
+     * Tests that when the DB server is down, the login timeout forces a return from {@link DatabaseFactory#getConnection}
+     * in a timely fashion (with an Exception).
+     */
+    @Test
+    public void testLoginTimeout() {
+        final Future<DatabaseEngine> dbEngineFuture = executor.submit(() -> DatabaseFactory.getConnection(dbProps));
+
+        assertThatCode(() -> dbEngineFuture.get(LOGIN_TIMEOUT_SECONDS + TIMEOUT_TOLERANCE_SECONDS, TimeUnit.SECONDS))
+                .as("When the DB server is down, the DB engine creation should fail by timeout in the connection")
+                .isInstanceOf(ExecutionException.class)
+                .hasCauseInstanceOf(DatabaseFactoryException.class);
+    }
+
+    /**
+     * Tests that after a connection with the DB server has been successfully established, if that connection is broken
+     * a given action returns in a timely fashion (with an Exception), according to the configured socket timeout.
+     *
+     * In this test the action performed is {@link DatabaseEngine#checkConnection()}.
+     * The connection between the test DB engine and the DB server is broken <strong>without being closed</strong>
+     * (this may happen for example when a load balancer changes the connections to a DB server without closing the old
+     * ones to the client).
+     */
+    @Test
+    public void testNetworkTimeout() throws Exception {
+        testRouter.init();
+
+        final DatabaseEngine dbEngine = executor.submit(() -> DatabaseFactory.getConnection(dbProps))
+                .get(LOGIN_TIMEOUT_SECONDS + TIMEOUT_TOLERANCE_SECONDS, TimeUnit.SECONDS);
+
+        Future<Boolean> connCheckFuture = executor.submit(() -> dbEngine.checkConnection());
+        assertTrue("PDB should be connected to the DB server", connCheckFuture.get(LOGIN_TIMEOUT_SECONDS, TimeUnit.SECONDS));
+
+        testRouter.breakConnections();
+
+        connCheckFuture = executor.submit(() -> dbEngine.checkConnection());
+
+        /*
+         we want to make sure that the login timeout (which is usually lower than socket timeout)
+          is not being used for timing out already established connections
+         */
+        TimeUnit.SECONDS.sleep(LOGIN_TIMEOUT_SECONDS + TIMEOUT_TOLERANCE_SECONDS);
+        assertThat(connCheckFuture)
+                .as("Connection check should only timeout after at least %d seconds", SOCKET_TIMEOUT_SECONDS)
+                .isNotDone();
+
+        assertFalse("After breaking connection, PDB should detect that it is not connected to the DB server",
+                connCheckFuture.get(SOCKET_TIMEOUT_SECONDS - LOGIN_TIMEOUT_SECONDS, TimeUnit.SECONDS));
+    }
+
+    /**
+     * A class that acts as a router, to forwards data sent between a {@link DatabaseEngine} in this test and the DB.
+     * This offers the possibility of interrupting the connections without closing them.
+     */
+    private static class TestRouter implements Closeable {
```

Comment:
This is good enough for this test but I don't think it would work very well for long lived connections (it doesn't have retry in case of read/write failures).

I guess if any test in the future needs this, we can extract it into its own class and improve it if needed...